File: conformance/python/contacts.py
"""Conformance consumer: contacts sample, Python target.

Exercises the generated ctypes wrapper end to end: the `ContactBook`
interface (its `__init__` calls the C constructor and `__del__` the destroy
symbol), enum marshalling, opaque struct handles with property getters,
optional strings, list-of-struct returns, boolean returns, and the typed
`ContactsError` subclasses (`InvalidName`, `NotFound`) raised by throwing
methods. The generated module is placed on sys.path via WV_PY; the cdylib is
selected with WEAVEFFI_LIBRARY.
"""
import os
import sys

sys.path.insert(0, os.environ["WV_PY"])

import contacts as wv  # noqa: E402


def main() -> None:
    book = wv.ContactBook()
    assert book.count() == 0

    alice = book.add("Alice", "Smith", "alice@example.com", wv.ContactType.Work)
    assert alice.id > 0
    assert alice.first_name == "Alice"
    assert alice.last_name == "Smith"
    assert alice.email == "alice@example.com"
    assert alice.contact_type == wv.ContactType.Work

    # Optional string: a missing email round-trips as None.
    bob = book.add("Bob", "Jones", None, wv.ContactType.Personal)
    fetched = book.get(bob.id)
    assert fetched.email is None
    assert fetched.contact_type == wv.ContactType.Personal

    assert book.count() == 2
    everyone = book.list()
    assert len(everyone) == 2
    assert {p.first_name for p in everyone} == {"Alice", "Bob"}

    # Typed error: an empty name is rejected with the InvalidName class of
    # the ContactsError domain.
    try:
        book.add("", "Smith", None, wv.ContactType.Personal)
        raise AssertionError("expected InvalidName for empty first name")
    except wv.ContactsError.InvalidName as exc:
        assert exc.code == 1, exc.code
        assert isinstance(exc, wv.ContactsError)
        assert isinstance(exc, wv.WeaveFFIError)
    assert book.count() == 2

    assert book.remove(alice.id) is True
    assert book.remove(alice.id) is False
    assert book.count() == 1

    # Typed error: a missing id raises NotFound; the bare name is the same
    # class as the scoped alias.
    try:
        book.get(9999)
        raise AssertionError("expected NotFound for missing contact")
    except wv.NotFound as exc:
        assert exc.code == 2, exc.code
        assert isinstance(exc, wv.ContactsError)
    assert wv.NotFound is wv.ContactsError.NotFound

    # A second book is independent state: contacts don't leak across objects.
    other = wv.ContactBook()
    assert other.count() == 0

    print("python/contacts: OK")


main()

File: conformance/ruby/contacts.rb
# frozen_string_literal: true
# Conformance consumer: contacts sample, Ruby target.
#
# Drives the 0.5.0 interface surface: ContactBook is a generated Ruby class
# wrapping the owned C object (released through FFI::AutoPointer), `new` maps
# to initialize, and methods pass the handle as the leading C argument.
# Throwing methods raise the typed ContactsError subclasses (InvalidName=1,
# NotFound=2); non-throwing methods keep the generic WeaveFFI::Error for
# panics only. The cdylib is selected via WEAVEFFI_LIBRARY.

$LOAD_PATH.unshift(File.join(ENV.fetch("WV_RB"), "lib"))
require "contacts"

def expect(cond, msg)
  raise "assertion failed: #{msg}" unless cond
end

book = WeaveFFI::ContactBook.new

alice = book.add("Alice", "Smith", "alice@example.com", WeaveFFI::ContactType::WORK)
expect(alice.id.positive?, "alice id positive")
expect(alice.first_name == "Alice", "first_name")
expect(alice.last_name == "Smith", "last_name")
expect(alice.email == "alice@example.com", "email")
expect(alice.contact_type == WeaveFFI::ContactType::WORK, "contact_type")

# Optional string: a missing email round-trips as nil.
bob = book.add("Bob", "Jones", nil, WeaveFFI::ContactType::PERSONAL)
expect(book.get(bob.id).email.nil?, "bob email nil")

expect(book.count == 2, "count == 2")
everyone = book.list
expect(everyone.length == 2, "list length == 2")
expect(everyone.map(&:first_name).sort == %w[Alice Bob], "list names")

expect(book.remove(alice.id) == true, "remove returns true")
expect(book.count == 1, "count == 1 after remove")
expect(book.remove(alice.id) == false, "second remove returns false")

# Typed errors: each domain code raises its own subclass of the domain
# class, which itself subclasses the generic error.
begin
  book.add("", "Smith", nil, WeaveFFI::ContactType::PERSONAL)
  raise "expected ContactsError::InvalidName for empty name"
rescue WeaveFFI::ContactsError::InvalidName => e
  expect(e.code == 1, "InvalidName code == 1 (got #{e.code})")
end

begin
  book.get(9999)
  raise "expected ContactsError::NotFound for missing contact"
rescue WeaveFFI::ContactsError::NotFound => e
  expect(e.code == 2, "NotFound code == 2 (got #{e.code})")
  expect(e.is_a?(WeaveFFI::ContactsError), "NotFound is a ContactsError")
  expect(e.is_a?(WeaveFFI::Error), "domain errors subclass WeaveFFI::Error")
end

# Rescuing the domain base class catches any code in the domain.
begin
  book.get(9999)
  raise "expected ContactsError"
rescue WeaveFFI::ContactsError => e
  expect(e.code == 2, "domain rescue sees NotFound code (got #{e.code})")
end

# Each book owns independent state; explicit destroy releases it early (the
# AutoPointer's GC release is then a no-op, so no double-free).
other = WeaveFFI::ContactBook.new
expect(other.count.zero?, "fresh book empty")
other.destroy

puts "ruby/contacts: OK"
